jsonrpc/http.go

```diff
@@ -53,32 +53,54 @@ func (h *HTTP) Run(ctx context.Context) error {
 	return <-errCh
 }
 
-// ServeHTTP processes an incoming HTTP request
-func (h *HTTP) ServeHTTP(writer http.ResponseWriter, req *http.Request) {
+type httpConn struct {
+	req *http.Request
+	rw  http.ResponseWriter
+}
+
+func accept(rw http.ResponseWriter, req *http.Request) *httpConn {
 	if req.Method == "GET" {
 		status := http.StatusNotFound
 		if req.URL.Path == "/" {
 			status = http.StatusOK
 		}
-		writer.WriteHeader(status)
-		return
+		rw.WriteHeader(status)
+		return nil
 	} else if req.Method != "POST" {
-		writer.WriteHeader(http.StatusMethodNotAllowed)
-		return
+		rw.WriteHeader(http.StatusMethodNotAllowed)
+		return nil
 	}
-
-	req.Body = http.MaxBytesReader(writer, req.Body, MaxRequestBodySize)
-	resp, err := h.rpc.HandleReader(req.Body)
-	writer.Header().Set("Content-Type", "application/json")
-	if err != nil {
-		writer.WriteHeader(http.StatusInternalServerError)
-	} else {
-		writer.WriteHeader(http.StatusOK)
+	return &httpConn{
+		req: req,
+		rw:  rw,
 	}
-	if resp != nil {
-		_, err = writer.Write(resp)
+}
+
+func (c *httpConn) Read(p []byte) (int, error) {
+	c.req.Body = http.MaxBytesReader(c.rw, c.req.Body, MaxRequestBodySize)
+	return c.req.Body.Read(p)
+}
+
+// Write returns the number of bytes of p written, not including the header.
+func (c *httpConn) Write(p []byte) (int, error) {
+	c.rw.Header().Set("Content-Type", "application/json")
+	c.rw.WriteHeader(http.StatusOK)
```

Comment:
dont we need a way to write non-200 status codes?

Reply:
Not here. If we've made it this far, we will only send a 200. `rw.Write` will do it automatically even if we don't.